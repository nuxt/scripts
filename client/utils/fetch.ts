export async function fetchScript(url: string) {
  const compressedResponse = await fetch(url, { headers: { 'Accept-Encoding': 'gzip' } }).catch((err) => {
    return {
      size: null,
      error: err,
    }
  })
  if (compressedResponse?.error) {
    return compressedResponse as { size: null, error: Error }
  }
  if (!compressedResponse.ok) {
    return {
      size: null,
      error: new Error(`Failed to fetch ${compressedResponse.status} ${compressedResponse.statusText}`),
    }
  }
  const size = await getResponseSize(compressedResponse)
  if (!size) {
    return {
      size: null,
    }
  }
  return {
    size: bytesToSize(size),
  }
}

async function getResponseSize(response: Response) {
  const reader = response.body?.getReader()
  const contentLength = response.headers.get('Content-Length')

  if (contentLength) {
    return Number(contentLength)
  }
  if (!reader) {
    return null
  }
  let total = 0
  let done = false
  while (!done) {
    const data = await reader.read()
    done = data.done
    total += data.value?.length || 0
  }
  return total > 0 ? total : null
}

function bytesToSize(bytes: number) {
  // be precise to 2 decimal places
  const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB']
  if (bytes === 0)
    return '0 Byte'
  const i = Math.floor(Math.log(bytes) / Math.log(1024))
  return `${Number.parseFloat((bytes / 1024 ** i).toFixed(2))} ${sizes[i]}`
}
