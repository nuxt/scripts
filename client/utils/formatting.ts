import { isRelative } from 'ufo'

export function humanFriendlyTimestamp(timestamp: number) {
  // use Intl.DateTimeFormat to format the timestamp, we only need the time aspect
  return new Intl.DateTimeFormat('en-US', {
    hour: 'numeric',
    minute: 'numeric',
    second: 'numeric',
    hour12: true,
  }).format(timestamp)
}

export function urlToOrigin(url: string) {
  if (isRelative(url, { acceptRelative: true }))
    return new URL(url).origin
  return url
}

export function msToHumanReadable(ms: number) {
  if (ms < 1000) {
    return ms + 'ms'
  }
  if (ms < 60000) {
    return (ms / 1000).toFixed(2) + 's'
  }
  return (ms / 60000).toFixed(2) + 'm'
}
