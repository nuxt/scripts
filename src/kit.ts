import { logger, tryUseNuxt } from '@nuxt/kit'
import { resolvePackageJSON } from 'pkg-types'
import { isCI, provider } from 'std-env'

const isStackblitz = provider === 'stackblitz'

interface EnsurePackageInstalledOptions {
  rootDir: string
  searchPaths?: string[]
  prompt?: boolean
}

async function promptToInstall(name: string, installCommand: () => Promise<void>, options: EnsurePackageInstalledOptions) {
  if (await resolvePackageJSON(name).catch(() => null))
    return true

  logger.info(`Package ${name} is missing`)
  if (isCI)
    return false

  // In StackBlitz we install packages automatically by default
  if (options.prompt === true || (options.prompt !== false && !isStackblitz)) {
    const confirm = await logger.prompt(`Do you want to install ${name} package?`, {
      type: 'confirm',
      name: 'confirm',
      initial: true,
    })

    if (!confirm)
      return false
  }

  logger.info(`Installing ${name}...`)
  try {
    await installCommand()
    logger.success(`Installed ${name}`)
    return true
  }
  catch (err) {
    logger.error(err)
    return false
  }
}

// TODO: refactor to Nuxi
const installPrompts = new Set<string>()
export function installNuxtModule(name: string, options?: EnsurePackageInstalledOptions) {
  if (installPrompts.has(name))
    return
  installPrompts.add(name)
  const nuxt = tryUseNuxt()
  if (!nuxt)
    return
  return promptToInstall(name, async () => {
    const { runCommand } = await import(String('nuxi'))
    await runCommand('module', ['add', name, '--cwd', nuxt.options.rootDir])
  }, { rootDir: nuxt.options.rootDir, searchPaths: nuxt.options.modulesDir, ...options })
}
