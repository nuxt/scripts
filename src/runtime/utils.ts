import { defu } from 'defu'
import type { GenericSchema, InferInput, ObjectSchema, ValiError } from 'valibot'
import type { UseScriptInput } from '@unhead/vue'
import { useScript } from './composables/useScript'
import { parse } from '#nuxt-scripts-validator'
import { useRuntimeConfig } from '#imports'
import type {
  EmptyOptionsSchema,
  NuxtUseScriptOptions,
  RegistryScriptInput,
  ScriptRegistry,
} from '#nuxt-scripts'

export type MaybePromise<T> = Promise<T> | T

function validateScriptInputSchema<T extends GenericSchema>(key: string, schema: T, options?: InferInput<T>) {
  if (import.meta.dev) {
    try {
      parse(schema, options)
    }
    catch (_e) {
      const e = _e as ValiError<any>
      // TODO nicer error handling
      // @ts-expect-error untyped?
      console.error(e.issues.map(i => `${key}.${i.path?.map(i => i.key).join(',')}: ${i.message}`).join('\n'))
    }
  }
}

type OptionsFn<O extends ObjectSchema<any, any>> = (options: InferInput<O>) => ({
  scriptInput?: UseScriptInput
  scriptOptions?: NuxtUseScriptOptions
  schema?: O
  clientInit?: () => void
})

export function scriptRuntimeConfig<T extends keyof ScriptRegistry>(key: T) {
  return ((useRuntimeConfig().public.scripts || {}) as ScriptRegistry)[key]
}

export function useRegistryScript<T extends Record<string | symbol, any>, O extends ObjectSchema<any, any> = EmptyOptionsSchema, U = {}>(registryKey: keyof ScriptRegistry | string, optionsFn: OptionsFn<O>, _userOptions?: RegistryScriptInput<O>) {
  const scriptConfig = scriptRuntimeConfig(registryKey as keyof ScriptRegistry)
  const userOptions = Object.assign(_userOptions || {}, typeof scriptConfig === 'object' ? scriptConfig : {})
  const options = optionsFn(userOptions)

  const scriptInput = defu(userOptions.scriptInput, options.scriptInput, { key: registryKey }) as any as UseScriptInput
  const scriptOptions = Object.assign(userOptions?.scriptOptions || {}, options.scriptOptions || {})
  if (import.meta.dev) {
    scriptOptions.devtools = defu(scriptOptions.devtools, { registryKey })
    if (options.schema) {
      const registryMeta: Record<string, string> = {}
      for (const k in options.schema.entries) {
        if (options.schema.entries[k].type !== 'optional') {
          registryMeta[k] = String(userOptions[k as any as keyof typeof userOptions])
        }
      }
      scriptOptions.devtools.registryMeta = registryMeta
    }
  }
  const init = scriptOptions.beforeInit
  scriptOptions.beforeInit = () => {
    // a manual trigger also means it was disabled by nuxt.config
    if (import.meta.dev && !scriptOptions.skipValidation && options.schema) {
      // overriding the src will skip validation
      if (!userOptions.scriptInput?.src) {
        validateScriptInputSchema(registryKey, options.schema, userOptions)
      }
    }
    // avoid clearing the user beforeInit
    init?.()
    if (import.meta.client) {
      // validate input in dev
      options.clientInit?.()
    }
  }
  return useScript<T, U>(scriptInput, scriptOptions as NuxtUseScriptOptions<T, U>)
}
