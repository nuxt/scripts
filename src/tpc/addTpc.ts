import { GooglaAnalyticsData, GoogleAnalytics, GoogleTagManager, GoogleTagManagerData, type Output } from 'third-party-capital'
import { addImports, addTemplate, useNuxt } from '@nuxt/kit'
import type { RegistryScript } from '../runtime/types'
import { extendTypes } from '../kit'
import { generateTpcContent, generateTpcTypes, type ScriptContentOpts } from './utils'

interface TpcDescriptor {
  input: ScriptContentOpts
  filename: string
  registry: RegistryScript
}

const scripts: TpcDescriptor[] = [{
  input: {
    data: GoogleTagManagerData as Output,
    scriptFunctionName: 'useScriptGoogleTagManager',
    use: () => {
      return { dataLayer: window.dataLayer, google_tag_manager: window.google_tag_manager }
    },
    stub: ({ fn }) => {
      return fn === 'dataLayer' ? [] : undefined
    },
    tpcKey: 'gtm',
    tpcTypeImport: 'GoogleTagManagerApi',
    featureDetectionName: 'nuxt-third-parties-gtm',
  },
  filename: 'nuxt-scripts-gtm',
  registry: {
    label: 'Google Tag Manager',
    category: 'tracking',
    logo: `<svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" viewBox="0 0 256 256"><path fill="#8AB4F8" d="m150.262 245.516l-44.437-43.331l95.433-97.454l46.007 45.091z"/><path fill="#4285F4" d="M150.45 53.938L106.176 8.731L9.36 104.629c-12.48 12.48-12.48 32.713 0 45.207l95.36 95.986l45.09-42.182l-72.654-76.407z"/><path fill="#8AB4F8" d="m246.625 105.37l-96-96c-12.494-12.494-32.756-12.494-45.25 0c-12.495 12.495-12.495 32.757 0 45.252l96 96c12.494 12.494 32.756 12.494 45.25 0c12.495-12.495 12.495-32.757 0-45.251"/><circle cx="127.265" cy="224.731" r="31.273" fill="#246FDB"/></svg>`,
    scriptBundling(options) {
      const data = GoogleTagManager(options)
      const mainScript = data.scripts?.find(({ key }) => key === 'gtag')

      if (mainScript && 'url' in mainScript && mainScript.url)
        return mainScript.url

      return false
    },
  },
}, {
  input: {
    data: GooglaAnalyticsData as Output,
    scriptFunctionName: 'useScriptGoogleAnalytics',
    use: () => {
      return { dataLayer: window.dataLayer, gtag: window.gtag }
    },
    // allow dataLayer to be accessed on the server
    stub: ({ fn }) => {
      return fn === 'dataLayer' ? [] : undefined
    },
    tpcKey: 'gtag',
    tpcTypeImport: 'GoogleAnalyticsApi',
    featureDetectionName: 'nuxt-third-parties-ga',
  },
  filename: 'nuxt-scripts-ga',
  registry: {
    label: 'Google Analytics',
    category: 'tracking',
    logo: `<svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" viewBox="0 0 256 256"><path fill="#8AB4F8" d="m150.262 245.516l-44.437-43.331l95.433-97.454l46.007 45.091z"/><path fill="#4285F4" d="M150.45 53.938L106.176 8.731L9.36 104.629c-12.48 12.48-12.48 32.713 0 45.207l95.36 95.986l45.09-42.182l-72.654-76.407z"/><path fill="#8AB4F8" d="m246.625 105.37l-96-96c-12.494-12.494-32.756-12.494-45.25 0c-12.495 12.495-12.495 32.757 0 45.252l96 96c12.494 12.494 32.756 12.494 45.25 0c12.495-12.495 12.495-32.757 0-45.251"/><circle cx="127.265" cy="224.731" r="31.273" fill="#246FDB"/></svg>`,
    scriptBundling(options) {
      const data = GoogleAnalytics(options)
      const mainScript = data.scripts?.find(({ key }) => key === 'gtag')

      if (mainScript && 'url' in mainScript && mainScript.url)
        return mainScript.url

      return false
    },
  },
}]

export function addTpc(registry: RegistryScript[]) {
  const nuxt = useNuxt()
  const fileImport = new Map<string, string>()

  for (const script of scripts) {
    extendTypes(script.filename, async () => await generateTpcTypes(script.input))

    const { dst } = addTemplate({
      getContents() {
        return generateTpcContent(script.input)
      },
      filename: `modules/${script.filename}.ts`,
    })

    addImports({
      from: dst,
      name: 'useScriptGoogleAnalytics',
    })

    script.registry.import = {
      from: dst,
      name: script.input.scriptFunctionName,
    }

    fileImport.set(script.filename, script.input.scriptFunctionName)

    registry.push(script.registry)
  }

  nuxt.hook('prepare:types', async ({ declarations }) => {
    declarations.push(`
${Array.from(fileImport).map(([filename, fn]) => `import { ${fn} } from '#build/modules/${filename}'`).join('\n')}
declare module '#imports' {
${Array.from(fileImport).map(([_, fn]) => `  export { ${fn} }`).join('\n')}
}    
    `)
  })
}
