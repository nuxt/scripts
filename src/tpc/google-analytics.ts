import { addImports, addTemplate, useNuxt } from '@nuxt/kit'
import type { Output } from 'third-party-capital'
import { GooglaAnalyticsData, GoogleAnalytics } from 'third-party-capital'
import type { RegistryScript } from '../runtime/types'
import { getTpcScriptContent } from './utils'

export default function googleAnalitycsRegistry() {
  const nuxt = useNuxt()
  const { dst } = addTemplate({
    getContents() {
      return getTpcScriptContent({
        data: GooglaAnalyticsData as Output,
        scriptFunctionName: 'useScriptGoogleAnalytics',
        use: () => {
          return { dataLayer: window.dataLayer, gtag: window.gtag }
        },
        // allow dataLayer to be accessed on the server
        stub: ({ fn }) => {
          return fn === 'dataLayer' ? [] : undefined
        },
        tpcKey: 'gtag',
        tpcTypeImport: 'GoogleAnalyticsApi',
      })
    },
    filename: 'nuxt-scripts/tpc/google-analytics.ts',
    write: true,
  })

  addImports({
    from: dst,
    name: 'useScriptGoogleAnalytics',
  })
  const registry: RegistryScript = {
    label: 'Google Tag Manager',
    category: 'tracking',
    logo: `<svg xmlns="http://www.w3.org/2000/svg" width="32" height="32" viewBox="0 0 256 256"><path fill="#8AB4F8" d="m150.262 245.516l-44.437-43.331l95.433-97.454l46.007 45.091z"/><path fill="#4285F4" d="M150.45 53.938L106.176 8.731L9.36 104.629c-12.48 12.48-12.48 32.713 0 45.207l95.36 95.986l45.09-42.182l-72.654-76.407z"/><path fill="#8AB4F8" d="m246.625 105.37l-96-96c-12.494-12.494-32.756-12.494-45.25 0c-12.495 12.495-12.495 32.757 0 45.252l96 96c12.494 12.494 32.756 12.494 45.25 0c12.495-12.495 12.495-32.757 0-45.251"/><circle cx="127.265" cy="224.731" r="31.273" fill="#246FDB"/></svg>`,
    import: {
      name: 'useScriptGoogleAnalytics',
      from: dst,
    },
    scriptBundling(options) {
      const data = GoogleAnalytics(options)
      const mainScript = data.scripts?.find(({ key }) => key === 'gtag')

      if (mainScript && 'url' in mainScript && mainScript.url)
        return mainScript.url

      return false
    },
  }
  nuxt.hook('scripts:registry', (scripts) => {
    scripts.push(registry)
  })
}
