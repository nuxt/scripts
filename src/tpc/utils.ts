import type { ExternalScript, Output, Script } from 'third-party-capital'
import { genImport, genTypeImport } from 'knitwork'
import { useNuxt } from '@nuxt/kit'
import type { HeadEntryOptions } from '@unhead/vue'

export interface ScriptContentOpts {
  data: Output
  scriptFunctionName: string
  tpcTypeImport: string
  tpcKey: string
  /**
   * This will be stringified. The function must be pure.
   */
  use: () => any
  /**
   * This will be stringified. The function must be pure.
   */
  stub: (params: { fn: string }) => any
}

const HEAD_VAR = '__head'
const INJECTHEAD_CODE = `const ${HEAD_VAR} =  injectHead()`

export function getTpcScriptContent(input: ScriptContentOpts) {
  const nuxt = useNuxt()
  if (!input.data.scripts)
    throw new Error('input.data has no scripts !')

  const mainScript = input.data.scripts?.find(({ key }) => key === input.tpcKey) as ExternalScript

  if (!mainScript)
    throw new Error(`no main script found for ${input.tpcKey} in third-party-capital`)

  const mainScriptOptions = getScriptInputOption(mainScript)

  const imports = new Set<string>([
    'import { withQuery } from \'ufo\'',
    'import { useRegistryScript } from \'#nuxt-scripts-utils\'',
    'import type { RegistryScriptInput } from \'#nuxt-scripts\'',
  ])

  const chunks: string[] = []
  const functionBody: string[] = []

  const hasParams = mainScript.params?.length

  if (input.tpcTypeImport) {
    // TPC type import
    imports.add(genTypeImport('third-party-capital', [input.tpcTypeImport]))
  }

  if (hasParams) {
    imports.add(genImport('#nuxt-scripts-validator', ['object', 'any']))
    // need schema validation from tpc
    chunks.push(`const OptionSchema = object({${mainScript.params?.map(p => `${p}:  any()`)}})`)
  }

  chunks.push(`
declare global {
  interface Window extends ${input.tpcTypeImport} {}
}`)

  const clientInitCode: string[] = []

  if (input.data.stylesheets) {
    if (!functionBody.includes(INJECTHEAD_CODE)) {
      functionBody.unshift(INJECTHEAD_CODE)
    }
    functionBody.push(`${HEAD_VAR}.push({link: ${JSON.stringify(input.data.stylesheets.map(s => ({ ref: 'stylesheet', href: s })))}})`)
  }

  for (const script of input.data.scripts) {
    if ('code' in script)
      clientInitCode.push(replaceTokenToRuntime(script.code))

    if (script === mainScript)
      continue

    if ('url' in script) {
      functionBody.push(`${HEAD_VAR}.push({scripts:{ async: true, src: ${script.url} }},${JSON.stringify(getScriptInputOption(script))})`)
    }
  }

  chunks.push(`export type Input = RegistryScriptInput${hasParams ? '<typeof OptionSchema>' : ''}`)

  chunks.push(`
export function ${input.scriptFunctionName}<T extends ${input.tpcTypeImport}>(_options?: Input) {
${functionBody.join('\n')}
  return useRegistryScript${hasParams ? '<T, typeof OptionSchema>' : ''}('${input.tpcKey}', options => ({
        scriptInput: {
            src: withQuery('${mainScript.url}', {${mainScript.params?.map(p => `${p}: options?.${p}`)}})
        },
        ${nuxt.options.dev ? 'schema: OptionSchema,' : ''}
        scriptOptions: {
            use: ${input.use.toString()},
            stub: import.meta.client ? undefined :  ${input.stub.toString()},
            ${mainScriptOptions ? `...(${JSON.stringify(mainScriptOptions)})` : ''}
        },
        ${clientInitCode.length ? `clientInit: import.meta.server ? undefined : () => {${clientInitCode.join('\n')}},` : ''}
    }), _options)
}`)

  chunks.unshift(...Array.from(imports))
  return chunks.join('\n')
}

function replaceTokenToRuntime(code: string) {
  return code.split(';').map(c => c.replaceAll(/'?\{\{(.*?)\}\}'?/g, 'options.$1')).join(';')
}

function getScriptInputOption(script: Script): HeadEntryOptions | undefined {
  if (script.location === 'body') {
    if (script.action === 'append') {
      return { tagPosition: 'bodyClose' }
    }
    return { tagPosition: 'bodyOpen' }
  }

  if (script.action === 'append') {
    return { tagPriority: 1 }
  }
}
