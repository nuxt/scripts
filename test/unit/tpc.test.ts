import { beforeEach, describe, expect, it, vi } from 'vitest'
import { useNuxt } from '@nuxt/kit'
import { TSESTree, parse } from '@typescript-eslint/typescript-estree'
import { generateTpcContent, type ScriptContentOpts } from '../../src/tpc/utils'

vi.mock('@nuxt/kit', async (og) => {
  const mod = await og<typeof import('@nuxt/kit')>()
  return {
    ...mod,
    useNuxt: vi.fn(mod.useNuxt),
  }
})

describe.each([
  {
    env: 'production',
    isDev: false,
  },
  {
    env: 'development',
    isDev: true,
  },
])('tpc composable generation in $env', ({ isDev }) => {
  beforeEach(() => {
    // @ts-expect-error mock only needed properties
    vi.mocked(useNuxt).mockReturnValue({ options: { dev: isDev } })
  })

  it('expect to throw if no main scripts', () => {
    expect(() => generateTpcContent({
      data: {
        scripts: [],
        id: 'google-analytics',
        description: 'for test purpose',
      },
      tpcKey: 'google-analytics',
      tpcTypeImport: 'GoogleAnalyticsInput',
      scriptFunctionName: 'useScriptGoogleAnalytics',
      use: () => { },
      stub: () => { },
    })).rejects.toThrow('no main script found for google-analytics in third-party-capital')
  })

  describe('script content generation', () => {
    const input: ScriptContentOpts = {
      data: {
        id: 'google-analytics',
        scripts: [
          {
            key: 'google-analytics',
            params: ['id'],
            url: 'https://www.google-analytics.com/analytics.js',
            strategy: 'client',
            location: 'head',
            action: 'prepend',
          },
        ],
        description: 'for test purpose',
      },
      tpcKey: 'google-analytics',
      tpcTypeImport: 'GoogleAnalyticsInput',
      scriptFunctionName: 'useScriptGoogleAnalytics',
      use: () => {
        return { dataLayer: window.dataLayer, gtag: window.gtag }
      },
      stub: () => {
        return []
      },
    }

    it(`expect to${isDev ? '' : ' not'} add the schema to the script options`, async () => {
      const result = await generateTpcContent(input)
      const returnStatement = getTpcScriptReturnStatement(result, 'useScriptGoogleAnalytics')
      if (!returnStatement || returnStatement.argument?.type !== TSESTree.AST_NODE_TYPES.CallExpression || (returnStatement.argument?.callee as TSESTree.Identifier).name !== 'useRegistryScript') {
        throw new Error('TPC Scripts must return a call expression of useRegistryScript')
      }
      const optionFnTree = returnStatement.argument.arguments[1] as TSESTree.ArrowFunctionExpression
      const optionFnReturn = optionFnTree.body as TSESTree.ObjectExpression
      const scriptOptionAst = optionFnReturn.properties.find((node): node is TSESTree.Property => node.type === TSESTree.AST_NODE_TYPES.Property && node.key.type === TSESTree.AST_NODE_TYPES.Identifier && node.key.name === 'scriptOptions')!

      const useFn = (scriptOptionAst.value as TSESTree.ObjectExpression).properties.find(node => node.type === TSESTree.AST_NODE_TYPES.Property && node.key.type === TSESTree.AST_NODE_TYPES.Identifier && node.key.name === 'use')
      if (!useFn) throw new Error('use function not found')
      expect(getCodeFromAst(result, useFn)).toContain('return { dataLayer: window.dataLayer, gtag: window.gtag }')

      const schemaNode = optionFnReturn.properties.find(node => node.type === TSESTree.AST_NODE_TYPES.Property && node.key.type === TSESTree.AST_NODE_TYPES.Identifier && node.key.name === 'schema')
      if (isDev) {
        expect(schemaNode).toBeTruthy()
        expect(getCodeFromAst(result, schemaNode!)).toContain('schema: OptionSchema')
      }
      else {
        expect(schemaNode).toBeUndefined()
      }
    })

    it('expect to stringify the use and stub functions', async () => {
      const result = await generateTpcContent(input)
      const returnStatement = getTpcScriptReturnStatement(result, 'useScriptGoogleAnalytics')
      if (!returnStatement || returnStatement.argument?.type !== TSESTree.AST_NODE_TYPES.CallExpression || (returnStatement.argument?.callee as TSESTree.Identifier).name !== 'useRegistryScript') {
        throw new Error('TPC Scripts must return a call expression of useRegistryScript')
      }
      const optionFnTree = returnStatement.argument.arguments[1] as TSESTree.ArrowFunctionExpression
      const optionFnReturn = optionFnTree.body as TSESTree.ObjectExpression
      const scriptOptionAst = optionFnReturn.properties.find((node): node is TSESTree.Property => node.type === TSESTree.AST_NODE_TYPES.Property && node.key.type === TSESTree.AST_NODE_TYPES.Identifier && node.key.name === 'scriptOptions')!

      const useFn = (scriptOptionAst.value as TSESTree.ObjectExpression).properties.find(node => node.type === TSESTree.AST_NODE_TYPES.Property && node.key.type === TSESTree.AST_NODE_TYPES.Identifier && node.key.name === 'use')
      if (!useFn) throw new Error('use function not found')
      expect(getCodeFromAst(result, useFn)).toContain('return { dataLayer: window.dataLayer, gtag: window.gtag }')

      const stubFn = (scriptOptionAst.value as TSESTree.ObjectExpression).properties.find(node => node.type === TSESTree.AST_NODE_TYPES.Property && node.key.type === TSESTree.AST_NODE_TYPES.Identifier && node.key.name === 'stub')
      if (!stubFn) throw new Error('stub function not found')
      expect(getCodeFromAst(result, stubFn)).toContain('return []')
    })

    it('expect to augment window types', async () => {
      const result = await generateTpcContent(input)
      const ast = parse(result, { loc: true, range: true })
      const augmentWindowTypes = ast.body.find((node): node is TSESTree.TSModuleDeclaration => node.type === TSESTree.AST_NODE_TYPES.TSModuleDeclaration)
      expect(augmentWindowTypes).toBeTruthy()
      expect(getCodeFromAst(result, augmentWindowTypes!)).toContain('interface Window extends GoogleAnalyticsInput {}')
    })
  })
})

describe('script content generation with head positioning', () => {
  const inputBase: ScriptContentOpts = {
    data: {
      id: 'google-analytics',
      scripts: [
        {
          key: 'google-analytics',
          params: ['id'],
          url: 'https://www.google-analytics.com/analytics.js',
          strategy: 'client',
          location: 'body',
          action: 'append',
        },
      ],
      description: 'for test purpose',
    },
    tpcKey: 'google-analytics',
    tpcTypeImport: 'GoogleAnalyticsInput',
    scriptFunctionName: 'useScriptGoogleAnalytics',
    use: () => { },
    stub: () => { },
  }

  describe('main script', () => {
    it('main script post body position', async () => {
      const scriptOptsAst = getTpcScriptOptsASt(await generateTpcContent(inputBase), 'useScriptGoogleAnalytics')
      expect(getCodeFromAst(await generateTpcContent(inputBase), scriptOptsAst)).toContain('"tagPosition":"bodyClose"')
    })
    it('main script pre body position', async () => {
      const scriptOptsAst = getTpcScriptOptsASt(await generateTpcContent({
        ...inputBase,
        data: {
          ...inputBase.data,
          scripts: [{
            ...inputBase.data.scripts![0],
            action: 'prepend',
          }],
        },
      }), 'useScriptGoogleAnalytics')
      expect(getCodeFromAst(await generateTpcContent(inputBase), scriptOptsAst)).toContain('"tagPosition":"bodyClose"')
    })
  })
})

function getTpcScriptAst(code: string, name: string) {
  const ast = parse(code, { loc: true, range: true })
  const tpcScriptAst = ast.body.find((node): node is TSESTree.ExportDefaultDeclaration => node.type === TSESTree.AST_NODE_TYPES.ExportNamedDeclaration && node.declaration?.type === TSESTree.AST_NODE_TYPES.FunctionDeclaration && node.declaration.id?.name === name)
  if (!tpcScriptAst) {
    throw new Error(`no function declaration found for ${name}`)
  }

  const functionAst = tpcScriptAst.declaration

  return functionAst as TSESTree.FunctionDeclaration
}

function getCodeFromAst(code: string, ast: TSESTree.Node) {
  return code.slice(ast.range[0], ast.range[1])
}

function getTpcScriptReturnStatement(code: string, name: string) {
  const tpcScriptAst = getTpcScriptAst(code, name)
  const returnStatement = tpcScriptAst.body.body.find((node): node is TSESTree.ReturnStatement => node.type === TSESTree.AST_NODE_TYPES.ReturnStatement)
  if (!returnStatement) {
    throw new Error('TPC Scripts must return a call expression of useRegistryScript')
  }
  return returnStatement
}

function getTpcScriptOptsASt(code: string, name: string) {
  const returnStatement = getTpcScriptReturnStatement(code, name)
  if (!returnStatement || returnStatement.argument?.type !== TSESTree.AST_NODE_TYPES.CallExpression || (returnStatement.argument?.callee as TSESTree.Identifier).name !== 'useRegistryScript') {
    throw new Error('TPC Scripts must return a call expression of useRegistryScript')
  }
  const optionFnTree = returnStatement.argument.arguments[1] as TSESTree.ArrowFunctionExpression
  const optionFnReturn = optionFnTree.body as TSESTree.ObjectExpression
  const scriptOptionAst = optionFnReturn.properties.find((node): node is TSESTree.Property => node.type === TSESTree.AST_NODE_TYPES.Property && node.key.type === TSESTree.AST_NODE_TYPES.Identifier && node.key.name === 'scriptOptions')!
  if (!scriptOptionAst) throw new Error('scriptOptions not found')
  return scriptOptionAst
}
