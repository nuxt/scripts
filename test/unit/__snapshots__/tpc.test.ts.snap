// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tpc composable generation in 'development' > expect to generate script content > composable generation in development 1`] = `
"import { withQuery } from 'ufo'
import { useRegistryScript } from '#nuxt-scripts-utils'
import type { RegistryScriptInput } from '#nuxt-scripts'
import type { GoogleAnalyticsInput } from "third-party-capital";
import { object, any } from "#nuxt-scripts-validator";
const OptionSchema = object({id:  any()})

declare global {
  interface Window extends GoogleAnalyticsInput {}
}
export type Input = RegistryScriptInput<typeof OptionSchema>

export function useScriptGoogleAnalytics<T extends GoogleAnalyticsInput>(_options?: Input) {

  return useRegistryScript<T, typeof OptionSchema>('google-analytics', options => ({
        scriptInput: {
            src: withQuery('https://www.google-analytics.com/analytics.js', {id: options?.id})
        },
        schema: OptionSchema,
        scriptOptions: {
            use: () => {
      },
            stub: true ? undefined :  () => {
      }
        },
        
    }), _options)
}"
`;

exports[`tpc composable generation in 'production' > expect to generate script content > composable generation in production 1`] = `
"import { withQuery } from 'ufo'
import { useRegistryScript } from '#nuxt-scripts-utils'
import type { RegistryScriptInput } from '#nuxt-scripts'
import type { GoogleAnalyticsInput } from "third-party-capital";
import { object, any } from "#nuxt-scripts-validator";
const OptionSchema = object({id:  any()})

declare global {
  interface Window extends GoogleAnalyticsInput {}
}
export type Input = RegistryScriptInput<typeof OptionSchema>

export function useScriptGoogleAnalytics<T extends GoogleAnalyticsInput>(_options?: Input) {

  return useRegistryScript<T, typeof OptionSchema>('google-analytics', options => ({
        scriptInput: {
            src: withQuery('https://www.google-analytics.com/analytics.js', {id: options?.id})
        },
        schema: OptionSchema,
        scriptOptions: {
            use: () => {
      },
            stub: true ? undefined :  () => {
      }
        },
        
    }), _options)
}"
`;
